src/map/mod.rs

```diff
@@ -0,0 +1,201 @@
+    core::{MessageResult, Mut, View, ViewId, ViewMarker},
+    DomNode, DynMessage, ElementProps, Pod, ViewCtx, WithStyle, HTML_NS,
+};
+
+mod events;
+pub use self::events::*;
+
+// TODO:
+// How can we pass tuples, like `()`, `(C)`, `(C, C)`?
+pub fn map<State, Action, C>(children: Vec<C>) -> Map<State, Action, C>
+where
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+    Map {
+        zoom: None,
+        center: None,
+        children,
+        phantom: PhantomData,
+    }
+}
+
+/// This is a marker trait that is used to pass only map-specific children
+/// to the [`map`] function.
+pub trait MapChild<State, Action>: View<State, Action, ViewCtx, DynMessage>
+where
+    Self::ViewState: MapChildViewState,
+{
+}
+
+/// This is a hack that allows children of the map
+/// to get access to [`leaflet::Map`].
+pub trait MapChildViewState {
+    fn after_build(&mut self, map: &leaflet::Map);
+    fn after_rebuild(&mut self, map: &leaflet::Map);
+}
+
+pub struct Map<State, Action, C>
+where
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+    children: Vec<C>,
+    zoom: Option<f64>,
+    center: Option<(f64, f64)>,
+    phantom: PhantomData<fn() -> (State, Action)>,
+}
+
+impl<State, Action, C> Map<State, Action, C>
+where
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+    pub fn zoom(mut self, value: f64) -> Self {
+        self.zoom = Some(value);
+        self
+    }
+
+    pub fn on_zoom_end<F>(&self, callback: F) -> OnZoomEnd<State, Action, F>
+    where
+        F: Fn(&mut State, f64) + 'static,
+    {
+        on_zoom_end(callback)
+    }
+
+    pub fn center(mut self, lat: f64, lng: f64) -> Self {
+        self.center = Some((lat, lng));
+        self
+    }
+}
+
+impl<State, Action, C> ViewMarker for Map<State, Action, C>
+where
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+}
+
+pub struct MapViewState<State, Action, C>
+where
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+    leaflet_map: leaflet::Map,
+    children: Vec<(C::Element, C::ViewState)>,
+}
+
+#[derive(Debug)]
+enum MapMessage {
+    InitMap,
+}
+
+impl<State, Action, C> View<State, Action, ViewCtx, DynMessage> for Map<State, Action, C>
+where
+    State: 'static,
+    Action: 'static,
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+    type Element = Pod<web_sys::HtmlElement, ElementProps>;
+
+    type ViewState = MapViewState<State, Action, C>;
+
+    fn build(&self, ctx: &mut ViewCtx) -> (Self::Element, Self::ViewState) {
+        // TODO:
+        // Is there a way to just use `html::div(())`?
+        let mut pod = Pod::new_element(vec![], HTML_NS, "div");
+        pod.props
+            .styles()
+            .set_style("width".into(), Some("100%".into()));
+        pod.props
+            .styles()
+            .set_style("height".into(), Some("100%".into()));
+        pod.node.apply_props(&mut pod.props);
```

Comment:
Hmm I just cleaned up yesterday the docs for and the modifiers (e.g. to fix that `after_build`), which will result in a PR soon, I'm actually waiting for the hydration PR being merged as I'm to lazy to cherry pick it from that branch to main.

`WithStyle` was the only modifier without docs...
I hope it's clear [how it's used](https://github.com/Philipp-M/xilem/blob/c1d7deac1c9493e8368611149f6869544913db40/xilem_web/src/style.rs#L108) now.

Or in other words, this is a state machine, and the marking methods need to be invoked in the correct order, otherwise that state machine may break...

As `Map` is an `impl DomView`, `apply_props` shouldn't be invoked, this happens automatically. It was only necessary to do this in `after_build` etc. because that callback should rely on the current state of the DOM element.

So in your case with the that branch above applied, it would be:

```suggestion
        let mut pod = Pod::new_element(vec![], HTML_NS, "div");
        pod.set_style("width".into(), Some("100%".into()));
        pod.set_style("height".into(), Some("100%".into()));
        pod.mark_end_of_style_modifier();
```

and as described in the docs, it would be necessary to do this in rebuild as well (+ `rebuild_style_modifier`). I think the best kind of docs in action would be the `kurbo_shape` views. Diffing and applying happens somewhere else (in `Styles::apply_style_changes`).


Anyway